src/commands/run.py

```diff
@@ -90,42 +95,51 @@ def family_runner(callstack, _a, _o, nodes):
         execute_threaded_batches(batches)
 
     def execute_threaded_batches(batches):
-        class JobRunner:
-            def __init__(self, job):
-                self.unsafe_job = job # This Job object may not thread safe
-                self.thread = threading.Thread(target=self.run)
+        loop = asyncio.get_event_loop()
+        def handler_interrupt():
+            print('Interrupt Received! ')
+            print('Cancelling the jobs...')
+            for task in asyncio.Task.all_tasks(loop = loop):
+                task.cancel()
+        loop.add_signal_handler(signal.SIGINT, handler_interrupt)
 
-            def run(self):
-                local_session = Session()
-                try:
-                    job = local_session.merge(self.unsafe_job)
-                    job.run()
-                    if job.exit_code == 0:
-                        symbol = 'Pass'
-                    else:
-                        symbol = 'Failed: {}'.format(job.exit_code)
-                    click.echo('ID: {}, Node: {}, {}'.format(job.id, job.node, symbol))
-                finally:
-                    local_session.commit()
-                    Session.remove()
+        max_ssh = int(os.environ.setdefault('ADMINWARE_MAX_SSH', '100'))
+        start_delay = float(os.environ.setdefault('ADMINWARE_START_DELAY', '0.2'))
+        pool = concurrent.futures.ThreadPoolExecutor(max_workers = max_ssh)
+
+        async def start_tasks(tasks):
+            active_tasks = []
+            def remove_done_tasks():
+                for active_task in active_tasks:
+                    if active_task._state == 'FINISHED':
+                        active_tasks.remove(active_task)
+                        break
+            for task in tasks:
+                while len(active_tasks) > max_ssh:
+                    remove_done_tasks()
+                    await asyncio.sleep(0.01)
+                asyncio.ensure_future(task, loop = loop)
+                active_tasks.append(task)
+                print('Starting Job: {}'.format(task.node))
+                await(asyncio.sleep(start_delay))
+            print('Waiting for jobs to finish...')
```

Comment:
I think these `print` statements should be `click.echo` for consistency with the rest of the code base. That being said I'm fairly sure we had an IRL chat about this & I can't remember the outcome.

Reply:
This has been resolved in #155